src/Operators/vector_rotation_operators.jl

```diff
@@ -0,0 +1,157 @@
+
+@inline extrinsic_vector(i, j, k, grid::AbstractGrid, uᵢ, vᵢ) = 
+    getvalue(uᵢ, i, j, k, grid), getvalue(vᵢ, i, j, k, grid)
+
+# Intrinsic and extrinsic conversion for `OrthogonalSphericalShellGrid`s,
+# i.e. curvilinear grids defined on a sphere which are locally orthogonal.
+# If the coordinates match with the coordinates of a latitude-longitude grid
+# (i.e. globally orthogonal), these functions collapse to 
+# uₑ, vₑ, wₑ = uᵢ, vᵢ, wᵢ
+
+# 2D vectors
+@inline function intrinsic_vector(i, j, k, grid::OrthogonalSphericalShellGrid, uₑ, vₑ)
+
+    φᶜᶠᵃ₊ = φnode(i, j+1, 1, grid, Center(), Face(), Center())
+    φᶜᶠᵃ₋ = φnode(i,   j, 1, grid, Center(), Face(), Center())
+    Δyᶜᶜᵃ = Δyᶜᶜᶜ(i,   j, 1, grid)
+
+    # θᵢ is the rotation angle between intrinsic and extrinsic reference frame
+    Rcosθᵢ = deg2rad(φᶜᶠᵃ₊ - φᶜᶠᵃ₋) / Δyᶜᶜᵃ
+
+    φᶠᶜᵃ₊ = φnode(i+1, j, 1, grid, Face(), Center(), Center())
+    φᶠᶜᵃ₋ = φnode(i,   j, 1, grid, Face(), Center(), Center())
+    Δxᶜᶜᵃ = Δxᶜᶜᶜ(i,   j, 1, grid)
+
+    Rsinθᵢ = - deg2rad(φᶠᶜᵃ₊ - φᶠᶜᵃ₋) / Δxᶜᶜᵃ
+
+    # Normalization for the rotation angles
+    Rᵢ = sqrt(Rcosθᵢ^2 + Rsinθᵢ^2)
+
+    u  = getvalue(uₑ, i, j, k, grid)
+    v  = getvalue(vₑ, i, j, k, grid)
+
+    cosθᵢ = Rsinθᵢ / Rᵢ
+    sinθᵢ = Rsinθᵢ / Rᵢ
+
+    uᵢ = u * cosθᵢ + v * sinθᵢ
+    vᵢ = u * sinθᵢ - v * cosθᵢ
+
+    return uᵢ, vᵢ
+end
+
+# 3D vectors
+@inline function intrinsic_vector(i, j, k, grid::OrthogonalSphericalShellGrid, uₑ, vₑ, wₑ)
+
+    uᵢ, vᵢ = intrinsic_vector(i, j, k, grid, uₑ, vₑ)
+    wᵢ = getvalue(wₑ, i, j, k, grid)
+
+    return uᵢ, vᵢ, wᵢ
+end
+
+# 2D vectors
+@inline function extrinsic_vector(i, j, k, grid::OrthogonalSphericalShellGrid, uᵢ, vᵢ)
+
+    φᶜᶠᵃ₊ = φnode(i, j+1, 1, grid, Center(), Face(), Center())
+    φᶜᶠᵃ₋ = φnode(i,   j, 1, grid, Center(), Face(), Center())
+    Δyᶜᶜᵃ = Δyᶜᶜᶜ(i,   j, 1, grid)
+
+    # θₑ is the rotation angle between intrinsic and extrinsic reference frame
+    Rcosθₑ = deg2rad(φᶜᶠᵃ₊ - φᶜᶠᵃ₋) / Δyᶜᶜᵃ
+
+    φᶠᶜᵃ₊ = φnode(i+1, j, 1, grid, Face(), Center(), Center())
+    φᶠᶜᵃ₋ = φnode(i,   j, 1, grid, Face(), Center(), Center())
+    Δxᶜᶜᵃ = Δxᶜᶜᶜ(i,   j, 1, grid)
+
+    Rsinθₑ = - deg2rad(φᶠᶜᵃ₊ - φᶠᶜᵃ₋) / Δxᶜᶜᵃ
+
+    # Normalization for the rotation angles
+    Rₑ = sqrt(Rcosθₑ^2 + Rsinθₑ^2)
+
+    u  = getvalue(uᵢ, i, j, k, grid)
+    v  = getvalue(vᵢ, i, j, k, grid)
+
+    cosθₑ = Rsinθₑ / Rₑ
+    sinθₑ = Rsinθₑ / Rₑ
+
+    uₑ = u * cosθₑ - v * sinθₑ
+    vₑ = u * sinθₑ + v * cosθₑ
+
+    return uₑ, vₑ
+end
+
+# 3D vectors
+@inline function extrinsic_vector(i, j, k, grid::OrthogonalSphericalShellGrid, uᵢ, vᵢ, wᵢ)
+
+    uₑ, vₑ = intrinsic_vector(i, j, k, grid, uᵢ, vᵢ)
+    wₑ = getvalue(wᵢ, i, j, k, grid)
+
+    return uₑ, vₑ, wₑ
+end
+
+#####
+##### Component-wise conversion between reference frames
+#####
+
+@inline intrinsic_vector_x_component(i, j, k, grid::AbstractGrid, uₑ, vₑ, wₑ) = 
+    @inbounds intrinsic_vector(i, j, k, grid, uₑ, vₑ, wₑ)[1]
```

Comment:
where are these used? Why did you decide on `intrinsic_vector_x_component` rather than `x_intrinsic_vector` or `intrinsic_vector_x` both of which are more common in the Oceananigans code base? 

Why do we not implement `intrinsic_vector_x_component` and then construct `intrinsic_vector` from that? It seems more natural than to write `intrinsic_vector` for all three components and then break it down like this.

Reply:
these are just used in the tests, we probably it is better to remove them from the src and define them where they are used. I though `intrinsic_vector_x_component` makes more sense because `x_vector` or `vector_x` does not really tell me that it is the x-component (sounds more like a vector that has to do with the x direction).

Regarding the second question, the vector transformation requires all the components and naturally returns all the components in case a transformation is needed (if we use the thin shell approximation then only the horizonal ones are needed). So going component-wise will double (or triple) the amount of computation.

Reply:
Ok, but then what is the point of the function. I think `intrinsic_vector(i, j, k, grid, uₑ, vₑ, wₑ)[1]` is actually more clear than `intrinsic_vector_x_component`

Reply:
you are right, I needed a function with a `Number` return value to be used in the tests in combination with `KernelFunctionOperation`. I have moved this function outside the `src` and inside the tests